### PR TITLE
refactor: TypeScript-ize jobs module

### PR DESCRIPTION
**Pull request by MaikoTan**
This will be a big long-term project (；・∀・)

- [x] buff_tracker.ts
- [x] combo_tracker.ts
- [x] constants.ts
- [x] jobs.ts
- [x] jobs_config.ts
- [x] utils.ts

As this module references files such as `overlay_plugin_api`, `widget_list`, `timerbar`, etc, it must be merged after them.

### CODE REVIEW

**Comment by MaikoTan** [comment]
This will be a big long-term project (；・∀・)

- [ ] buff_tracker.ts
- [x] combo_tracker.ts
- [ ] constants.ts
- [ ] jobs.ts
- [ ] jobs_config.ts
- [ ] utils.ts

- components/
  - [ ] ast.ts
  - [ ] blm.ts
  - [ ] blu.ts
  - [ ] brd.ts
  - [ ] dnc.ts
  - [ ] drg.ts
  - [ ] drk.ts
  - [ ] gnb.ts
  - [ ] index.ts
  - [ ] mch.ts
  - [ ] mnk.ts
  - [ ] nin.ts
  - [ ] pld.ts
  - [ ] rdm.ts
  - [ ] sam.ts
  - [ ] sch.ts
  - [ ] smn.ts
  - [ ] war.ts
  - [ ] whm.ts

**Comment by quisquous** [comment]
Feel free to land these separately as you go in other PRs!

**Comment by MaikoTan** [comment]
> Feel free to land these separately as you go in other PRs!

The main reason is that the files are cross-referenced to each other and it is difficult for me to keep the module still functional in a separated PR. However, perhaps we can separate the job-specific components.

**Comment by MaikoTan** [comment]
Would continue working on when `resources/widgetlist.ts` landed.

File: ui/jobs/jobs.ts [hunk 1]
```diff
@@ -1002,31 +1089,32 @@ class Bars {
...
+    logs.push(t + '1A:10000000:' + me + ' gains the effect of Brotherhood from Other Guy for 15.0 Seconds.');
+    const e = { type: 'onLogEvent' as const, detail: { logs: logs } };
     this._onLogEvent(e);
   }
 }
 
-let gBars;
+let gBars: Bars;
 
 UserConfig.getUserConfigLocation('jobs', Options, () => {
```
Comment (MaikoTan) [inline ui/jobs/jobs.ts:1119]: @quisquous I am not very understand how was this function working, can you help me, basically for typing the argument `Options`?

Comment (quisquous) [reply ui/jobs/jobs.ts:1119]: My thought was that `const Options` up at the top of the file will become `type JobsOptions extends BaseOptions`.  And then this type here is `JobsOptions`.

**Comment by quisquous** [comment]
Is it possible to land any of this separately? Like the buff tracker or the combo tracker on its own?

**Comment by MaikoTan** [comment]
> Is it possible to land any of this separately? Like the buff tracker or the combo tracker on its own?

I'll try, but there are some exports cross-referenced by each other and not easy to separate them.

**Comment by MaikoTan** [comment]
Separated in multiple PRs so close this one. (╯°□°）╯︵ ┻━┻

**Pull request closed without merging**